File: Python3/206. Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        # if not head or not head.next: # 44ms
        #     return head
        # a, b, c = None, head, head.next
        # while c:
        #     b.next = a
        #     a, b, c = b, c, c.next
        # b.next = a
        # return b
            
        if not head: # 48ms
            return head
        prev, node = None, head
        while node:
            tmp = node.next
            node.next = prev
            prev = node
            node = tmp
        return prev

File: Golang/206. Reverse Linked List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode { // 4ms
    if head == nil {
        return head
    } else {
        var prev *ListNode
        node := head
        for node != nil {
            tmp := node.Next
            node.Next = prev
            prev = node
            node = tmp
        }
        return prev
    }
}
